### ISSUE TITLE
nltk.parse.generate.generate only works with finite language grammars

### ISSUE THREAD

**Issue by ghost** [nltk/nltk#7]
Generate will run out of stack on recursive grammars. It also crashes on a certain edge case, but I forget what it was.

I had to write a grammar once, and it would have been useful for testing to generate some sentences from it.

**Comment by alexrudnick** [comment]
Thanks dlila! Would you include a test grammar that definitely makes it run out of stack?
(a quick demonstration in the repl would be really helpful.)

Let us know if you find the edge case :)

**Comment by ghost** [comment]
Any grammar that defines a language with an infinite number of strings makes it run out of stack.
For example
    A -> A B | 'a'
    B -> 'b'

The edge case I mentioned was an unrelated problem. I will try to find it and will post it if I do. I think it was an index out of bounds error caused by the [0] in the last line of "generate".

Anyway, I posted a patch on nltk-dev that should fix both problems.

[EDIT]: ... but I just got an e-mail saying that my e-mail was not delivered. Do I need to join nltk-dev to post patches on it?
[EDIT2]: I guess I do need to be a member to submit patches. I submitted a membership request.

**Comment by stevenbird** [comment]
Thanks for the contribution.  Would you mind adding some in-line documentation please, and attaching a diff which does not use CRLF newlines?  http://help.github.com/line-endings/

**Comment by ghost** [comment]
Hi.

Sorry about the new lines. I should have checked the patch before sending.

I have attached a new one, with comments, some simplifications, and no
CRLFs.

Regards,
Denis.

On Wed, Dec 28, 2011 at 12:05 AM, Steven Bird <
reply@reply.github.com

**Comment by stevenbird** [comment]
Where is the attachment please?

**Comment by ghost** [comment]
Sorry about that. I have put the patch up on dropbox, now:
http://dl.dropbox.com/u/29228818/patch

On Sun, Jan 8, 2012 at 3:41 AM, Steven Bird <
reply@reply.github.com

**Comment by alvations** [comment]
The `nltk.parse.generate.demo()` is a working example.

``` python
>>> from nltk.parse import generate
>>> generate.demo()
Generating the first 23 sentences for demo grammar:

  S -> NP VP
  NP -> Det N
  PP -> P NP
  VP -> 'slept' | 'saw' NP | 'walked' PP
  Det -> 'the' | 'a'
  N -> 'man' | 'park' | 'dog'
  P -> 'in' | 'with'

  1. the man slept
  2. the man saw the man
  3. the man saw the park
  4. the man saw the dog
  5. the man saw a man
  6. the man saw a park
  7. the man saw a dog
  8. the man walked in the man
  9. the man walked in the park
 10. the man walked in the dog
 11. the man walked in a man
 12. the man walked in a park
 13. the man walked in a dog
 14. the man walked with the man
 15. the man walked with the park
 16. the man walked with the dog
 17. the man walked with a man
 18. the man walked with a park
 19. the man walked with a dog
 20. the park slept
 21. the park saw the man
 22. the park saw the park
 23. the park saw the dog
```

But @ghost is right about the infinite loops:

``` python
>>> from nltk.parse import generate
>>> generate.demo_grammar = """
... A -> A B | 'a'
... B -> 'b'
... """
>>> generate.demo()
```

[out]:

```
Generating the first 23 sentences for demo grammar:

A -> A B | 'a'
B -> 'b'

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "nltk/parse/generate.py", line 72, in demo
    for n, sent in enumerate(generate(grammar, n=N), 1):
  File "nltk/parse/generate.py", line 41, in _generate_all
    for frag1 in _generate_one(grammar, items[0], depth):
  File "nltk/parse/generate.py", line 51, in _generate_one
    for frag in _generate_all(grammar, prod.rhs(), depth-1):
  File "nltk/parse/generate.py", line 41, in _generate_all
    for frag1 in _generate_one(grammar, items[0], depth):
  File "nltk/parse/generate.py", line 51, in _generate_one
    for frag in _generate_all(grammar, prod.rhs(), depth-1):
... (repeating errors)
  File "nltk/parse/generate.py", line 50, in _generate_one
    for prod in grammar.productions(lhs=item):
  File "nltk/grammar.py", line 558, in productions
    return self._lhs_index.get(lhs, [])
RuntimeError: maximum recursion depth exceeded
```

**Comment by alvations** [comment]
Within python, there is a default recursion limit that will kick in at runtime: https://docs.python.org/3.5/library/sys.html#sys.setrecursionlimit

So I think we can either close this issue and let the users find out the recursion nature of their CFG or add a mechanism to check for recursion at [grammar initialization](https://github.com/nltk/nltk/blob/develop/nltk/grammar.py#L421) or at [generation](https://github.com/nltk/nltk/blob/develop/nltk/parse/generate.py#L39). 

I'm not sure how we could check for "malicious" or "valid" recursion though.

We have the case of the infamous NN compound that yield the same infinite recursion:

```
noun_phrase_grammar = """
NP -> N NP  | "foo bar" 
N -> 'foo' | 'bar'
"""
```

And possibly, the "linguistic" solution would be to include the `NP -> N N` rule (but it has to be the first rule before `NP -> N NP`:

``` python
>>> generate.demo_grammar = """
... NP -> N N | N NP  | "foo bar"
... N -> "foo" | "bar"
... """
>>> generate.demo()
Generating the first 23 sentences for demo grammar:

NP -> N N | N NP  | "foo bar"
N -> "foo" | "bar"

  1. foo foo
  2. foo bar
  3. bar foo
  4. bar bar
  5. foo foo foo
  6. foo foo bar
  7. foo bar foo
  8. foo bar bar
  9. foo foo foo foo
 10. foo foo foo bar
 11. foo foo bar foo
 12. foo foo bar bar
 13. foo foo foo foo foo
 14. foo foo foo foo bar
 15. foo foo foo bar foo
 16. foo foo foo bar bar
 17. foo foo foo foo foo foo
 18. foo foo foo foo foo bar
 19. foo foo foo foo bar foo
 20. foo foo foo foo bar bar
 21. foo foo foo foo foo foo foo
 22. foo foo foo foo foo foo bar
 23. foo foo foo foo foo bar foo
```

**Comment by alvations** [comment]
Given that Python checks for recursion, we could do catch the error and warn the user with the following at https://github.com/nltk/nltk/blob/develop/nltk/parse/generate.py#L39 :

``` python
def _generate_all(grammar, items, depth):
    if items:
        try:
            for frag1 in _generate_one(grammar, items[0], depth):
                for frag2 in _generate_all(grammar, items[1:], depth):
                    yield frag1 + frag2
        except RuntimeError as _error:
            if _error.message == "maximum recursion depth exceeded":
                print("The grammar has rule(s) that yield infinite recursion!!\n", file=sys.stderr)
            raise RuntimeError

    else:
        yield []
```

E.g.

``` python
>>> from nltk.parse import generate
>>> generate.demo_grammar = """
... A -> A B | 'a'
... B -> 'b'
... """
>>> 
>>> generate.demo()
Generating the first 23 sentences for demo grammar:

A -> A B | 'a'
B -> 'b'

The grammar has rule(s) that yield infinite recursion!!

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "nltk/parse/generate.py", line 78, in demo
    for n, sent in enumerate(generate(grammar, n=N), 1):
  File "nltk/parse/generate.py", line 48, in _generate_all
    raise RuntimeError
RuntimeError
``
```

**Comment by fievelk** [comment]
@alvations @stevenbird Do you think we can close this issue?

**Comment by alvations** [comment]
Closing issue, solved on #1365 with raised warning
